### ISSUE TITLE
storage_service::join_node_response_handler assertion failure on retry

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#15956]
Seen locally with `topology.test_change_ip` and `topology_experimental_raft.test_fencing` in dev mode in a branch off of 6cc5bcae801d2e1298870acd3b3432fbb6aa3c57.

For example:
```
25,982 [shard 0: gms] storage_service - raft topology: coordinator accepted request to join, waiting for nodes {127.220.196.27} to be alive before responding and continuing
INFO  2023-11-05 20:58:25,983 [shard 0: gms] storage_service - raft topology: nodes {127.220.196.27} are alive
INFO  2023-11-05 20:58:25,983 [shard 0:strm] storage_service - raft topology: join: success
INFO  2023-11-05 20:58:25,983 [shard 0: gms] storage_service - raft topology: coordinator accepted request to join, waiting for nodes {127.220.196.27} to be alive before responding and continuing
INFO  2023-11-05 20:58:25,983 [shard 0: gms] storage_service - raft topology: nodes {127.220.196.27} are alive
scylla: seastar/include/seastar/core/future.hh:624: void seastar::future_state<seastar::internal::monostate>::set(A &&...) [T = seastar::internal::monostate, A = <>]: Assertion `_u.st == state::future' failed.
Aborting on shard 0.
Backtrace:
  /home/bhalevy/dev/scylla/build/dev/seastar/libseastar.so+0x3bfccc
  /home/bhalevy/dev/scylla/build/dev/seastar/libseastar.so+0x3ebe02
  /lib64/libc.so.6+0x3dbaf
  /lib64/libc.so.6+0x8e883
  /lib64/libc.so.6+0x3dafd
  /lib64/libc.so.6+0x2687e
  /lib64/libc.so.6+0x2679a
  /lib64/libc.so.6+0x36186
  0x24832ee
  0x24c3faa
  /home/bhalevy/dev/scylla/build/dev/seastar/libseastar.so+0x3ce96f
  /home/bhalevy/dev/scylla/build/dev/seastar/libseastar.so+0x3cfaf7
  /home/bhalevy/dev/scylla/build/dev/seastar/libseastar.so+0x3cefc9
  /home/bhalevy/dev/scylla/build/dev/seastar/libseastar.so+0x346d58
  /home/bhalevy/dev/scylla/build/dev/seastar/libseastar.so+0x345f92
  0xfe5729
  0x101e670
  0xfe4706
  /lib64/libc.so.6+0x27b89
  /lib64/libc.so.6+0x27c4a
  0xfe3264
```

```
(gdb) up
#5  0x00007ff914a58187 in __assert_fail (assertion=0x2fe4eb "_u.st == state::future", 
    file=0x3343a2 "seastar/include/seastar/core/future.hh", line=624, 
    function=0x39fe45 "void seastar::future_state<seastar::internal::monostate>::set(A &&...) [T = seastar::internal::monostate, A = <>]") at assert.c:101
101       __assert_fail_base (_("%s%s%s:%u: %s%sAssertion `%s' failed.\n%n"),
(gdb) up
#6  0x00000000024832ef in seastar::future_state<seastar::internal::monostate>::set<>() (
    this=0x6000010fbc70) at ./seastar/include/seastar/core/future.hh:624
624             assert(_u.st == state::future);
(gdb) p _u
$1 = {
  st = seastar::future_state_base::state::invalid,
  ex = {
    _M_exception_object = 0x0
  }
}
```

**Comment by bhalevy** [comment]
I think that what could have happened here is that the first call hit an error and called https://github.com/scylladb/scylladb/blob/6cc5bcae801d2e1298870acd3b3432fbb6aa3c57/service/storage_service.cc#L6689
Then, the exception was taken by a call to e.g. https://github.com/scylladb/scylladb/blob/6cc5bcae801d2e1298870acd3b3432fbb6aa3c57/service/storage_service.cc#L2656

And `future_state_base::take_exception`, sets the state as `invalid`, see https://github.com/scylladb/seastar/blob/37560d53a9d1dbcf79d217a8a85fe60a57fecf30/include/seastar/core/future.hh#L444-L457

The, on the retry, https://github.com/scylladb/scylladb/blob/6cc5bcae801d2e1298870acd3b3432fbb6aa3c57/service/storage_service.cc#L6606 doesn't see the shared_promise as `available()`, since `invalid` is not considered to be an available state.  See https://github.com/scylladb/seastar/blob/37560d53a9d1dbcf79d217a8a85fe60a57fecf30/include/seastar/core/future.hh#L440

**Comment by bhalevy** [comment]
Cc @gleb-cloudius @kbr-scylla

**Comment by gleb-cloudius** [comment]
@piodul can you look at this?

**Comment by bhalevy** [comment]
This looks like a seastar issue with shared_promise.
The following test case fails for me with scylladb/seastar@37560d53a9d1dbcf79d217a8a85fe60a57fecf30:
```c++
SEASTAR_THREAD_TEST_CASE(test_shared_future_available_with_exception) {
    shared_promise<int> spr;
    auto f0 = spr.get_shared_future();
    spr.set_exception(std::make_exception_ptr(std::runtime_error("expected error")));
    BOOST_REQUIRE(spr.available());
}
```

**Comment by bhalevy** [comment]
Same for:
```c++
SEASTAR_THREAD_TEST_CASE(test_shared_future_available_with_value) {
    shared_promise<int> spr;
    auto f0 = spr.get_shared_future();
    spr.set_value(42);
    BOOST_REQUIRE(spr.available());
}
```

**Comment by piodul** [comment]
Interestingly, inserting a yield between `set_value` and `available` causes the test to consistently pass:

```c++
SEASTAR_THREAD_TEST_CASE(test_shared_future_available_with_value) {
    shared_promise<int> spr;
    auto f0 = spr.get_shared_future();
    spr.set_value(42);
    BOOST_REQUIRE(spr.available());
}
```

This "fix" also works for the `test_shared_future_available_with_exception` test case.

**Comment by piodul** [comment]
@bhalevy I think your theory describes a legitimate issue.

The only missing detail is that `shared_state::resolve` is called _in a continuation_ which is attached to `_original_future`. The `_original_future` waits on a seastar promise inside `shared_promise`, and `shared_promise::set_value` sets that promise. However, setting a seastar promise does not cause the continuation attached to the future to run immediately, it is only `seastar::schedule`d to run later, at the end of the queue of the current scheduling group. Therefore, `available()` does not become true immediately after `set_value`.

If, after the task which calls `set_value`, another task which calls `available()` is already scheduled, then the other task will also see the shared future as unavailable.

Despite that, it's still not clear to me how exactly the events would have to play out in `join_node_response_handler` in order to trigger the issue.

**Comment by piodul** [comment]
@bhalevy you mentioned that you've hit the same failure twice, so it sounds like it should be possible to reproduce it. Do you still have it and could share it with me?

**Comment by bhalevy** [comment]
> @bhalevy you mentioned that you've hit the same failure twice, so it sounds like it should be possible to reproduce it. Do you still have it and could share it with me?

I have the core dumps and logs, but I didn't save the binary unfortunately.
I can try to reproduce...

**Comment by piodul** [comment]
I can try building and reproducing it, but you mentioned in the commit description that it was in a "branch off of https://github.com/scylladb/scylladb/commit/6cc5bcae801d2e1298870acd3b3432fbb6aa3c57". Could you share the branch?

**Comment by bhalevy** [comment]
@piodul, here it is: https://github.com/bhalevy/scylla/commits/branch-15956

**Comment by piodul** [comment]
> @piodul, here it is: https://github.com/bhalevy/scylla/commits/branch-15956

Thanks

**Comment by bhalevy** [comment]
Reproduced with the following change:
```diff
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 42cc479ba2..45c9f7225b 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -6777,14 +6777,17 @@ void storage_service::init_messaging_service(sharded<db::system_distributed_keys
                 return ss.join_node_request_handler(std::move(params));
             });
         });
-        ser::join_node_rpc_verbs::register_join_node_response(&_messaging.local(), [this] (raft::server_id dst_id, service::join_node_response_params params) {
-            return container().invoke_on(0, [dst_id, params = std::move(params)] (auto& ss) mutable -> future<join_node_response_result> {
+        ser::join_node_rpc_verbs::register_join_node_response(&_messaging.local(), [this] (raft::server_id dst_id, service::join_node_response_params params) -> future<join_node_response_result> {
+            auto func = [dst_id, params = std::move(params)] (auto& ss) mutable -> future<join_node_response_result> {
                 co_await ss._join_node_group0_started.get_shared_future(ss._abort_source);
                 if (ss._group0->load_my_id() != dst_id) {
                     throw raft_destination_id_not_correct(ss._group0->load_my_id(), dst_id);
                 }
-                co_return co_await ss.join_node_response_handler(std::move(params));
-            });
+                co_return co_await ss.join_node_response_handler(params);
+            };
+            auto ret = co_await container().invoke_on(0, func);
+            co_await container().invoke_on(0, func);
+            co_return ret;
         });
     }
 }
```

The following is actually enough too, but I wanted to simulated the rpc being called twice.
```diff
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 42cc479ba2..7134ae9cae 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -6783,7 +6783,9 @@ void storage_service::init_messaging_service(sharded<db::system_distributed_keys
                 if (ss._group0->load_my_id() != dst_id) {
                     throw raft_destination_id_not_correct(ss._group0->load_my_id(), dst_id);
                 }
-                co_return co_await ss.join_node_response_handler(std::move(params));
+                auto ret = co_await ss.join_node_response_handler(params);
+                co_await ss.join_node_response_handler(params);
+                co_return ret;
             });
         });
     }
```

**Comment by bhalevy** [comment]
`co_await yield()` after `_join_node_response_done.set_value()` doesn't help either.
I think it's related to the function the `overloaded_functor` being a coroutine.
OTOH, inserting a `co_await sleep(std::chrono::milliseconds(1))` does induce the required preemption and `_join_node_response_done.available()` is seen as `true` after `get_units(_join_node_response_handler_mutex, 1)` resolves in the queued call.

All in all, adding a `done` flag to be checked instead of the `shared_promise.available()` state seems to work well:
```diff
diff --git a/service/storage_service.hh b/service/storage_service.hh
index d779020a11..d2378eb4a1 100644
--- a/service/storage_service.hh
+++ b/service/storage_service.hh
@@ -819,6 +819,7 @@ class storage_service : public service::migration_listener, public gms::i_endpoi
     shared_promise<> _join_node_request_done;
     shared_promise<> _join_node_group0_started;
     shared_promise<> _join_node_response_done;
+    bool _join_node_response_done_flag = false;
     semaphore _join_node_response_handler_mutex{1};
 
     friend class join_node_rpc_handshaker;
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 42cc479ba2..71f5e2d0aa 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -6603,7 +6603,7 @@ future<join_node_response_result> storage_service::join_node_response_handler(jo
     // Wait until we sent and completed the join_node_request RPC
     co_await _join_node_request_done.get_shared_future(_abort_source);
 
-    if (_join_node_response_done.available()) {
+    if (_join_node_response_done_flag) {
         // We already handled this RPC. No need to retry it. Return immediately for idempotence.
         slogger.info("raft topology: the node got join_node_response RPC for the second time, ignoring");
         co_return join_node_response_result{};
@@ -6680,6 +6680,7 @@ future<join_node_response_result> storage_service::join_node_response_handler(jo
             // Unblock waiting join_node_rpc_handshaker::post_server_start,
             // which will start the raft server and continue
             _join_node_response_done.set_value();
+            _join_node_response_done_flag = true;
 
             co_return join_node_response_result{};
         },
@@ -6687,6 +6688,7 @@ future<join_node_response_result> storage_service::join_node_response_handler(jo
             auto eptr = std::make_exception_ptr(std::runtime_error(
                     format("the topology coordinator rejected request to join the cluster: {}", rej.reason)));
             _join_node_response_done.set_exception(std::move(eptr));
+            _join_node_response_done_flag = true;
 
             co_return join_node_response_result{};
         },
```

**Comment by piodul** [comment]
Ok, I see how it could happen. The code that waits for the shared future runs on the streaming scheduling group, while the handler that sets the shared promise runs on the gossiper scheduling group. The task created by `_original_future.then_wrapped` inherits the SG that was active at the time it was created, so `_join_node_response_done.set_value` sets the task in the non-current SG. It sounds plausible that the second RPC could happen very quickly and run before tasks from the streaming SG are executed, and would explain why the millisecond sleep helped.

I'll see how much work it would be to fix `seastar_promise`, if it's too much then I'll go with the `bool` for now. Thanks for the help.

**Comment by piodul** [comment]
I've run `test_change_ip` with modifications described [here](https://github.com/scylladb/scylladb/issues/15956#issuecomment-1796288087) and reproduced the failure. Then, after applying the seastar fix from https://github.com/scylladb/seastar/pull/1928 the crash no longer happens.

The fix was merged, I suppose that this issue can be closed after the seastar submodule is updated. cc: @scylladb/scylla-maint

**Comment by nyh** [comment]
> The fix was merged, I suppose that this issue can be closed after the seastar submodule is updated. cc: @scylladb/scylla-maint

Opened a PR to update the Seastar module: https://github.com/scylladb/scylladb/pull/15994
There are quite a few Seastar changes there, so let's see the CI passes.

**Comment by nyh** [comment]
Unfortunately, the Seastar patch had to be reverted, so this issue was **not** fixed.

**Comment by kbr-scylla** [comment]
Note: we're currently waiting on https://github.com/scylladb/seastar/pull/1958

**Comment by piodul** [comment]
A better fix was merged yesterday, I hope it won't get reverted now... @scylladb/scylla-maint can we try updating the seastar module again?
